crates/cli/src/suite.rs

```
@@ -0,0 +1,139 @@
+                .benchmarks
+                .iter()
+                .map(|p| p.display().to_string())
+                .collect(),
+            BenchmarkOrSuite::Benchmark(path) => vec![path.display().to_string()],
+        }
+    }
+}
+
+/// It is helpful to reference the original path.
+impl AsRef<OsStr> for BenchmarkOrSuite {
+    fn as_ref(&self) -> &OsStr {
+        match self {
+            BenchmarkOrSuite::Suite(suite) => suite.path.as_os_str(),
+            BenchmarkOrSuite::Benchmark(path) => path.as_os_str(),
+        }
+    }
+}
+
+/// Parse a [BenchmarkOrSuite] from a string path; files ending in `.suite` are
+/// assumed to be suite files.
+impl FromStr for BenchmarkOrSuite {
+    type Err = anyhow::Error;
+    fn from_str(path: &str) -> Result<Self, Self::Err> {
+        Ok(if Suite::has_extension(path) {
+            Self::Suite(Suite::parse(path)?)
+        } else {
+            Self::Benchmark(PathBuf::from(path))
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct Suite {
+    path: PathBuf,
+    benchmarks: Vec<PathBuf>,
+}
+
+impl Suite {
+    /// Parse the contents of a suite file:
+    /// - empty lines are ignored
+    /// - `#`-prefixed lines are ignored
+    /// - relative paths are resolved using the parent directory of the
+    ///   `suite_path`.
+    pub fn parse<P: AsRef<Path>>(suite_path: P) -> Result<Self> {
+        Self::parse_contents(suite_path.as_ref(), &fs::read(suite_path.as_ref())?)
+    }
+
+    /// Utility function for easier testing.
+    fn parse_contents<P: AsRef<Path>>(suite_path: P, file_contents: &[u8]) -> Result<Self> {
+        use io::BufRead;
+        let suite_path = suite_path.as_ref().to_path_buf();
+        let parent_dir = suite_path
+            .parent()
+            .expect("the suite path must have a parent directory");
+        let mut benchmarks = vec![];
+        for line in io::BufReader::new(file_contents)
+            .lines()
+            .filter_map(Result::ok)
```

This will silently ignore any errors. Instead we should do the (arguably uglier) thing where we `?`-propagate the error at the start of the loop body:

```rust
for line in io::BufReader::new(file_contents).lines() {
    let line = line?;
    // ...
}
```